tool/tsh/tsh.go

```diff
@@ -988,6 +997,91 @@ func refuseArgs(command string, args []string) {
 	}
 }
 
+// rawIdentity encodes the basic components of an identity file.
+type rawIdentity struct {
+	PrivateKey []byte
+	Certs      struct {
+		SSH []byte
+		TLS []byte
+	}
+	CACerts struct {
+		SSH [][]byte
+		TLS [][]byte
+	}
+}
+
+// decodeIdentity attempts to break up the contents of an identity file
+// into its respective components.
+func decodeIdentity(r io.Reader) (*rawIdentity, error) {
```

Comment:
Why didn't you go with `pem.Decode`? This function simplifies to:

```go
func parse(rest []byte) ([]*pem.Block, []byte) {
	var block *pem.Block
	var blocks []*pem.Block

	for {
		block, rest = pem.Decode(rest)
		if block == nil {
			return blocks, rest
		}
		blocks = append(blocks, block)
	}
}
```

You'll get a slice with the PEM blocks for the TLS key, certificate, and CA certificate and `[]byte` with the SSH user and host key.

Reply:
I may be missing something, but I don't believe that will work.  The `pem.Decode` function discards all data preceding the PEM block, so `rest` is just whatever happens to be *after* the last PEM block.  The identity file format doesn't guarantee that all PEM blocks come first.  In fact, there are tests in `tsh_test.go` which explicitly verify that `loadIdentity` functions correctly regardless of the relative order of the ssh certs and PEM block.

I looked at `pem.Decode` when I was originally writing this function.  This was my thinking at the time (let me know if you concur):

Given the lossy nature of `pem.Decode`, I would still need to scan line-by-line and execute control-flow based on line prefix.  Given that the PEM blocks are validated after being extracted (by functions which call `pem.Decode` internally), nothing would be added from a validation/correctness standpoint by using it.  This meant that the only benefit was removing the need to manually locate the end of the PEM block (which is just another prefix check).  Meanwhile, using `pem.Decode` would prevent me from using the very nice `bufio.Scanner` API, and would add an additional re-encoding step to get the extracted block back into the format expected by `client.Key`.  All in all, it seemed that using `pem.Decode` here would reduce readability without offering a tangible benefit.

Reply:
Okay, sounds good to me.